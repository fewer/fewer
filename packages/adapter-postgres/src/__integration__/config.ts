export default {
  database: 'postgres',
  user: 'postgres',
  password: 'password',
  host: 'postgres',
};
