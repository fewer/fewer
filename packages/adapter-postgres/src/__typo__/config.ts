export default {
  database: 'fewer_integration_tests',
  user: 'postgres',
  password: 'password',
  host: 'postgres',
};
