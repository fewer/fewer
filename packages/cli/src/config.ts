import cosmiconfig from 'cosmiconfig';

export interface FewerConfigurationFile {
  src: string;
  migrations: string;
  repositories: string;
  databases: string[];
  typescript: boolean;
  cjs: boolean;
}

const DEFAULT_CONFIG: FewerConfigurationFile = {
  src: 'src',
  migrations: 'src/migrations',
  repositories: 'src/repositories',
  databases: ['src/database.ts'],
  typescript: true,
  cjs: false,
};

let userConfig: FewerConfigurationFile | undefined;
const explorer = cosmiconfig('fewer');
const result = explorer.searchSync(process.cwd());

if (result) {
  // TODO: Should we validate the format here on runtime?
  userConfig = result.config as FewerConfigurationFile;
}

const config: FewerConfigurationFile = {
  ...DEFAULT_CONFIG,
  ...userConfig,
};

export default config;
