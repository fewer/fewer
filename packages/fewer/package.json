{
  "name": "fewer",
  "version": "0.0.6",
  "description": "A minimal ORM for Node.js.",
  "keywords": [
    "orm",
    "sql",
    "fewer"
  ],
  "author": "Jordan Gensler <jgensler@netflix.com>",
  "homepage": "https://github.com/fewer/fewer",
  "license": "MIT",
  "main": "lib/index.js",
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/fewer/fewer.git"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "bugs": {
    "url": "https://github.com/fewer/fewer/issues"
  },
  "dependencies": {
    "@fewer/sq": "^0.0.4"
  },
  "devDependencies": {
    "@fewer/typeval": "^0.0.4"
  }
}
