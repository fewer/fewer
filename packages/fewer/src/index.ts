import Schema from './Schema';
import { Repository, createRepository } from './Repository';

export { Schema, Repository, createRepository };
