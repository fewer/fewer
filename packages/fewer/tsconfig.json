{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "lib",
    "composite": true,
    "rootDir": "src"
  },
  "include": ["./src"],
  "references": [
    { "path": "../sq" },
  ]
}
