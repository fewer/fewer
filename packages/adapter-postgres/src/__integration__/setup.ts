import { createDatabase } from 'fewer';
import { Adapter } from '../';
import rawQuery from '../rawQuery';
import config from './config';

export async function prepare() {
  const adapter = new Adapter({...config, database: 'postgres'});

  const database = createDatabase({ adapter });

  try {
    await database.connect();
    await rawQuery(
      adapter.client!,
      'DROP DATABASE IF EXISTS fewer_integration_tests;',
    );
    await rawQuery(adapter.client!, 'CREATE DATABASE fewer_integration_tests;');
  } catch (err) {
    console.log(err);
  }

  await database.disconnect();
}

// prepare();
