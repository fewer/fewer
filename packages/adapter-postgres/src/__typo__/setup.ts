import { createDatabase, createMigration } from 'fewer';
import { Adapter, rawQuery } from '../';
import config from './config';

async function prepare() {
  const adapter = new Adapter({ ...config, database: 'postgres' });

  const pgDatabase = createDatabase({ adapter });

  const adapter2 = new Adapter({ ...config, database: 'fewer_typo_tests' });
  const database = createDatabase({ adapter: adapter2 });

  try {
    await pgDatabase.connect();
    await rawQuery(
      adapter.client!,
      'DROP DATABASE IF EXISTS fewer_typo_tests;',
    );
    await rawQuery(adapter.client!, 'CREATE DATABASE fewer_typo_tests;');

    await database.connect();

    const migration = createMigration(1, database, {
      change: (m, t) =>
        m
          .createTable(
            'users',
            { primaryKey: 'id' },
            {
              id: t.bigserial(),
              first_name: t.string(),
              last_name: t.string(),
            },
          )
          .createTable(
            'posts',
            { primaryKey: 'id' },
            {
              id: t.bigserial(),
              user_id: t.bigint(),
              title: t.string(),
              subtitle: t.string(),
            },
          ),
    });

    await migration.run('up');
  } catch (err) {
    console.log(err);
  }

  await database.disconnect();
  await pgDatabase.disconnect();
}

prepare();
