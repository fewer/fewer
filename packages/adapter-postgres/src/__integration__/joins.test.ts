
import { createSchema, createHasMany, createRepository, createBelongsTo, createMigration, createDatabase, Database } from 'fewer';
import { Adapter, rawQuery } from '..';
import config from './config';
import { prepare } from './setup';

type AdapterInstance = InstanceType<typeof Adapter>;

function getSchemaAndRepos(database: Database) {
  const schema = createSchema()
    .table(database, 'users', t => ({
      id: t.bigint(),
      first_name: t.string(),
      last_name: t.string(),
    }))
    .table(database, 'posts', t => ({
      title: t.string(),
      user_id: t.bigint(),
      subtitle: t.string(),
    }));

  const Users = createRepository(schema.tables.users);
  const Posts = createRepository(schema.tables.posts);

  const userPosts = createHasMany(Users, Posts, 'user_id');
  const belongsToUser = createBelongsTo(Users, 'user_id');

  return {
    schema, Users, Posts, userPosts, belongsToUser
  }
}

describe('load associations', () => {
  describe('postgres', () => {
    let database: Database<AdapterInstance>;
    let adapter: AdapterInstance;

    beforeEach(async () => {
      await prepare();
      adapter = new Adapter(config);
      database = createDatabase({ adapter });
      await database.connect();

      const migration = createMigration(1, database, {
        change: (m, t) =>
          m.createTable('users', null, {
            id: t.bigserial({ primaryKey: true }),
            first_name: t.string(),
            last_name: t.string(),
          }).createTable('posts', null, {
            id: t.bigserial({ primaryKey: true }),
            user_id: t.bigint(),
            title: t.string(),
            subtitle: t.string(),
          }),
      });

      await migration.run('up');
    });

    afterEach(async () => {
      await database.disconnect();
    });

    it('can join users <-> posts', async () => {
      const dbTypes = getSchemaAndRepos(database);

      const emilyId = await dbTypes.Users.create({
        first_name: 'Emily',
        last_name: 'Dobervich',
      });

      const jordanId = await dbTypes.Users.create({
        first_name: 'Jordan',
        last_name: 'Gensler',
      });

      const fooId = await dbTypes.Users.create({
        first_name: 'Foo',
        last_name: 'Bar',
      });

      await dbTypes.Posts.create({
        user_id: emilyId.id,
        title: 'How to use Fewer',
      });

      await dbTypes.Posts.create({
        user_id: jordanId.id,
        title: 'Abusing Typescript',
      });

      await dbTypes.Posts.create({
        user_id: jordanId.id,
        title: 'Ten Typescript Type Tricks',
      });

      const results = await dbTypes.Users.join('posts', dbTypes.userPosts);
      expect(results.sort((a, b) => a.id - b.id)).toMatchSnapshot();
    });
  });
});
