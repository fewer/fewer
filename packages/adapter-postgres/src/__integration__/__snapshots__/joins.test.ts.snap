// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`join associations postgres can join users <-> posts 1`] = `
Array [
  Object {
    "first_name": "Emily",
    "id": "1",
    "last_name": "Dobervich",
    "subtitle": null,
    "title": "How to use Fewer",
    "user_id": "1",
  },
  Object {
    "first_name": "Jordan",
    "id": "2",
    "last_name": "Gensler",
    "subtitle": null,
    "title": "Abusing Typescript",
    "user_id": "2",
  },
  Object {
    "first_name": "Jordan",
    "id": "3",
    "last_name": "Gensler",
    "subtitle": null,
    "title": "Ten Typescript Type Tricks",
    "user_id": "2",
  },
]
`;

exports[`join associations postgres nested joins 1`] = `
Array [
  Object {
    "first_name": "Emily",
    "id": "1",
    "last_name": "Dobervich",
    "subtitle": null,
    "title": "How to use Fewer",
    "user_id": "1",
  },
]
`;